blobs.go

```diff
@@ -0,0 +1,149 @@
+package distribution
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+)
+
+var (
+	// ErrBlobExists returned when blob already exists
+	ErrBlobExists = fmt.Errorf("blob exists")
+
+	// ErrBlobDigestUnsupported when blob digest is an unsupported version.
+	ErrBlobDigestUnsupported = fmt.Errorf("unsupported blob digest")
+
+	// ErrBlobUnknown when blob is not found.
+	ErrBlobUnknown = fmt.Errorf("unknown blob")
+
+	// ErrBlobUploadUnknown returned when upload is not found.
+	ErrBlobUploadUnknown = fmt.Errorf("blob upload unknown")
+
+	// ErrBlobInvalidLength returned when the blob has an expected length on
+	// commit, meaning mismatched with the descriptor or an invalid value.
+	ErrBlobInvalidLength = fmt.Errorf("blob invalid length")
+)
+
+// ErrBlobInvalidDigest returned when digest check fails.
+type ErrBlobInvalidDigest struct {
+	Digest digest.Digest
+	Reason error
+}
+
+func (err ErrBlobInvalidDigest) Error() string {
+	return fmt.Sprintf("invalid digest for referenced layer: %v, %v",
+		err.Digest, err.Reason)
+}
+
+// Descriptor describes targeted content. Used in conjunction with a blob
+// store, a descriptor can be used to fetch, store and target any kind of
+// blob. The struct also describes the wire protocol format. Fields should
+// only be added but never changed.
+type Descriptor struct {
+	// MediaType describe the type of the content. All text based formats are
+	// encoded as utf-8.
+	MediaType string `json:"mediaType,omitempty"`
+
+	// Length in bytes of content.
+	Length int64 `json:"length,omitempty"`
+
+	// Digest uniquely identifies the content. A byte stream can be verified
+	// against against this digest.
+	Digest digest.Digest `json:"digest,omitempty"`
+
+	// NOTE: Before adding a field here, please ensure that all
+	// other options have been exhausted. Much of the type relationships
+	// depend on the simplicity of this type.
+}
+
+// BlobDescriptorService makes blob descriptor data available by digest. The
+// service may provide a descriptor of a different digest if the provided
+// digest is not canonical.
+type BlobDescriptorService interface {
+	// Stat provides metadata about a blob identified by the digest.
+	Stat(ctx context.Context, dgst digest.Digest) (Descriptor, error)
```

Comment:
What do you think about changing stat to return existence? A lot of calls to Stat do a type check on the error for this.

e.g.
`Stat(ctx context.Context, dgst digest.Digest) (Descriptor, exists, error)`

this 
```
	desc, err := blobs.Stat(bh, bh.Digest)
	if err != nil {
		if err == distribution.ErrBlobUnknown {
			w.WriteHeader(http.StatusNotFound)
			bh.Errors.Push(v2.ErrorCodeBlobUnknown, bh.Digest)
		} else {
			bh.Errors.Push(v2.ErrorCodeUnknown, err)
		}
		return
	}
```
would become
```
	desc, exists, err := blobs.Stat(bh, bh.Digest)
	if err != nil {
		bh.Errors.Push(v2.ErrorCodeUnknown, err)
        return
	}
	if !exists {
		w.WriteHeader(http.StatusNotFound)
		bh.Errors.Push(v2.ErrorCodeBlobUnknown, bh.Digest)
	}
```